# (PART) Introduction {-}


# Epigenetic Data

## What is epigenetic data?
<br> As you may already know, typically all of the cells in your body contain the same DNA. How, then, do we have different cell types in our body? Your DNA contains a script that is able to produce the proteins required for each specific cell in your body. Which proteins, and subsequently which cells are made, depends on gene expression and regulation, i.e. "the way each cell deploys its genome."^[@ralston2008]

<br>***Epigenetic data*** arises from "the study of heritable and stable changes in gene expression that occur through alterations in the chromosome rather than in the DNA sequence."^[@nora2023] 


<br>
![Epigenetic Mechanisms](images/Epigenetic_Mechanisms.png "Epigenetic Mechanisms")
<div style="text-align: center;">
[commonfund.nih.gov](https://commonfund.nih.gov/sites/default/files/epigeneticmechanisms.pdf){target="_blank"}
</div>
<br>
The image above shows quite simply the basics of genetic structures. Several more complex processes are involved during cell replication such as DNA transcription and translation in order to make proteins. A key takeaway in coming closer to understanding gene expression is that **Chromatin** is a complex structure made up of DNA wound around histone proteins, with some segments of DNA being accessible/inaccessible to further processes. **Euchromatin** refers to the accessible state, while **Heterochromatin** refers to a chromatin state in which DNA cannot be transcribed (inaccessible).^[@shahid2023] There are many different epigenetic modifications that affect chromatin accessibility.

<br>
Some common epigenetic modifications include:

1. **DNA Methylation**: Addition of methyl groups to DNA, affecting gene expression regulation^[@nora2023].
2. **Histone Modifications**: Chemical changes to histone proteins that DNA wraps around, including acetylation, methylation, or phosphorylation. These changes influence chromatin structure and gene accessibility.^[@Kouzarides2007]
3. **Chromatin Accessibility**: Regions of open chromatin that are accessible to transcription factors (special types of proteins that bind to DNA sequences and regulate gene expression) further dictate which regions of DNA can be expressed^[@melanie2021].
<br>

In studying gene expression and epigenetic modifications, we can more closely understand biological mechanisms that regulate development, disease, and how cells respond to epigenetic factors.

<br>

### Common Epigenetic Techniques:
1.	***ATAC-Seq*** (Assay for Transposase-Accessible Chromatin with Sequencing):
<br><br>o```      Measures chromatin accessibility to identify open regions of the genome where transcription factors can bind.```
<br><br>o```      Output: Peaks indicating accessible chromatin regions.```
<br><br>

2. ***ChIP-Seq*** (Chromatin Immunoprecipitation Sequencing):
<br><br>o```      Used to identify DNA regions bound by specific proteins (e.g., transcription factors, histones with specific modifications).```
<br><br>o```      Output: Peaks indicating binding sites or modification locations.```

## What does epigenetic data look like?

Epigenetic data can be represented in various forms, depending on the type of modification being studied and the methods used to gather the data. **ATAC-Seq** and **ChIP-Seq** are the common methods I will focus on, but there are others that may produce different forms of data, such as WGS (whole-genome sequencing) which produces nucleotide sequencing data, or Bisulfite conversion of DNA producing data on methylation levels across the genome. 


### Representing epigenetic data
1.	***Raw Sequence Reads:***
<br><br>o```      These are the basic output of sequencing experiments, such as those from ChIP-Seq or ATAC-Seq.```
<br><br>o```      Reads are processed and aligned to a reference genome before undergoing peak calling.```
<br><br>
2.	***Peak Calling:***
<br><br>o```      A method used to identify regions in the genome where there is significant enrichment of sequencing reads. This indicates the presence of DNA-protein interactions (e.g., transcription factor binding sites) or accessible chromatin regions.```
<br><br>o```      Peaks represent areas where epigenetic marks or chromatin accessibility are concentrated.```
<br><br>

<div style="margin-left: 40px;">
***Representing Peaks:***
<br><br>
o ***P-value or Fold-change:***
<br><br>P-value: Indicates the statistical significance of the peak, helping to distinguish true peaks from background noise.
<br><br>Fold-change: Represents the difference in read density between treated and control samples, indicating the strength of the signal.
<br><br>
o ***Types of Peaks:***
<br><br>Categorical Peaks: Simple yes/no indication of a peakâ€™s presence.
<br><br>Continuous Peaks: More nuanced representation that includes the intensity or enrichment level of the peak, often visualized as a signal track.
<br><br>Thresholded/Pseudoreplicated Peaks: Usually categorical, these peaks are of high confidence regions from multiple replicates (experiments) or pseudoreplicates (artificial data splits), to ensure reliability and reproducibility.
</div>

<br>
<div style="text-align: center;">
***Example Data Pipeline***
</div>

![](images/chIP-Seq_pipeline.png "Example Data Pipeline")

<div style="text-align: center;">
[encodeproject.org](https://www.encodeproject.org/experiments/ENCSR817LUF/){target="_blank"}
</div>

<br>
This example data pipeline originates from a ChIP-seq experiment targeting the H3K36me3 histone modification in brain tissue. The aim of the experiement is to map the locations where the H3K36me3 histone modification is present along the genome. The yellow bubbles represent downloadable data sets of different types, while the blue boxes represent step types (e.g. peak calling). In the left column are multiple data sets of raw sequence reads, which then undergo data quality steps before being aligned (first blue box) to the reference human genome GRCh38 (denoted by ENCFF110MCL below the reads). The next steps include Peak calling (categorical peaks) and signal generation (continuous peaks) to produce the data we normally use in our machine learning models. This data pipeline process aids in normalisation, noise reduction, and dimensionality reduction of the data.

<br>

### Example Data Representations:

<br>
The following is an example of what this same chIP-Seq data targeting the H3K36me3 histone modification in brain tissue looks like using UCSC's Genome Browser. The pseudoreplicated peaks represent categorically, the significant locations along the genome where the H3K36me3 histone modification is present.
<br>

<iframe src="myhtml/ucsc_genome_browser.html" width="100%" height="510"></iframe>
<div style="text-align: center;">
</div>
<div style="text-align: center;">
[UCSC Genome Browser](https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&lastVirtModeType=default&lastVirtModeExtraState=&virtModeType=default&virtMode=0&nonVirtPosition=&position=chr1%3A11084744%2D11095920&hgsid=2307713234_Kap236Tjt6ZGnnNrXMkIhq2Ajn27){target="_blank"}
</div

<br><br><br><br><br><br><br>

<br>
The following is an example of ATAC-Seq data from an experiment on T-helper 17 cells (a type of immune system cell). Recall that the ATAC-Seq method aims to find chromatin regions that are accessible for transcription factor binding. The p-value and fold change graphs show continuous peaks, while the IDR thresholded peaks and pseudoreplicated peaks represent the significant locations of accessible chromatin along the genome.
<br>

<iframe src="myhtml/ucsc_gb_atacseq.html" width="100%" height="510"></iframe>
<div style="text-align: center;">
</div>
<div style="text-align: center;">
[UCSC Genome Browser](https://genome.ucsc.edu/cgi-bin/hgTracks?db=hg38&lastVirtModeType=default&lastVirtModeExtraState=&virtModeType=default&virtMode=0&nonVirtPosition=&position=chr1%3A88379533%2D113275174&hgsid=2307721306_mcnECXS4Hy0fNQ4yz3ZQTL7nimkW){target="_blank"}
</div

<br>

### Transformations to stop extreme p-values
When utlising genomic data which incorporates p-values, it is important to consider and deal with extreme p-values. One way this is done is through using an Arcsinh-transformation (inverse hyperbolic sine). 
<br><br>
$\text{arsinh}(x) = \ln \left( x + \sqrt{x^2 + 1} \right)$
<br>

The arcsinh-transformation as a logarithmic function helps in reducing the significance of outliers and sequencing depth while maintaining variance by compressing the range of the data. This transformation can be used in the data preprocessing stage.

## Sources of epigenetic data
There are numerous public data banks which contain genomic datasets ready to be downloaded.
<br><br>
[Blueprint](https://projects.ensembl.org/blueprint/){target="_blank"}
<br>
Blueprint's genomic datasets are focused on gene expression in healthy and diseased cells mostly relating to haematopoietic cells (cells which develop into different types of blood cells).
<br>

[Roadmap](https://www.ncbi.nlm.nih.gov/geo/roadmap/epigenomics/){target="_blank"}
<br>
The National Institute of Health's Roadmap Epigenomics Project contains sample datasets from multiple experiements as well as reference and mapping datasets. 
<br>

[Encode](https://www.encodeproject.org/){target="_blank"}
<br>
The Encode Project contains a large amount of publicly available genomic data easily filtered and downloaded. The genomic data used in this markdown book is sourced from Encode.
<br>

The largest genomic data bank is the [UK Biobank](https://www.ukbiobank.ac.uk/){target="_blank"}, however they require that you apply for access to their datasets.
<br>

## Handling bigWig files


## Data loaders and pre-processing
## Dealing with missing data (oversampling, undersampling, weighting)

