<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>3 gReLU | Machine Learning in Genomics: Containerised tutorials demonstrating best practises, pitfalls, and reproducibility</title>
  <meta name="description" content="a set of reproducible, containerized tutorials that include all necessary data, code, and descriptions to replicate key results, along with demonstrations of common pitfalls in the field of genomics." />
  <meta name="generator" content="bookdown 0.39 and GitBook 2.6.7" />

  <meta property="og:title" content="3 gReLU | Machine Learning in Genomics: Containerised tutorials demonstrating best practises, pitfalls, and reproducibility" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="a set of reproducible, containerized tutorials that include all necessary data, code, and descriptions to replicate key results, along with demonstrations of common pitfalls in the field of genomics." />
  <meta name="github-repo" content="https://github.com/sn2023imperial/genomicsmarkdown" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="3 gReLU | Machine Learning in Genomics: Containerised tutorials demonstrating best practises, pitfalls, and reproducibility" />
  
  <meta name="twitter:description" content="a set of reproducible, containerized tutorials that include all necessary data, code, and descriptions to replicate key results, along with demonstrations of common pitfalls in the field of genomics." />
  

<meta name="author" content=" Sach Nehal" />


<meta name="date" content="2024-09-03" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="loss-functions-and-peak-metrics.html"/>
<link rel="next" href="kipoi.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>



<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Machine Learning in Genomics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>About</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="epigenetic-data.html"><a href="epigenetic-data.html"><i class="fa fa-check"></i><b>1</b> Epigenetic Data</a>
<ul>
<li class="chapter" data-level="1.1" data-path="epigenetic-data.html"><a href="epigenetic-data.html#what-is-epigenetic-data"><i class="fa fa-check"></i><b>1.1</b> What is epigenetic data?</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="epigenetic-data.html"><a href="epigenetic-data.html#what-does-dna-look-like"><i class="fa fa-check"></i><b>1.1.1</b> What Does DNA Look Like?</a></li>
<li class="chapter" data-level="1.1.2" data-path="epigenetic-data.html"><a href="epigenetic-data.html#common-epigenetic-sequencing-techniques"><i class="fa fa-check"></i><b>1.1.2</b> Common Epigenetic Sequencing Techniques:</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="epigenetic-data.html"><a href="epigenetic-data.html#what-does-epigenetic-data-look-like"><i class="fa fa-check"></i><b>1.2</b> What does epigenetic data look like?</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="epigenetic-data.html"><a href="epigenetic-data.html#representing-epigenetic-data"><i class="fa fa-check"></i><b>1.2.1</b> Representing epigenetic data</a></li>
<li class="chapter" data-level="1.2.2" data-path="epigenetic-data.html"><a href="epigenetic-data.html#transformations-to-stop-extreme-p-values"><i class="fa fa-check"></i><b>1.2.2</b> Transformations to stop extreme p-values</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="epigenetic-data.html"><a href="epigenetic-data.html#sources-of-epigenetic-data"><i class="fa fa-check"></i><b>1.3</b> Sources of epigenetic data</a></li>
<li class="chapter" data-level="1.4" data-path="epigenetic-data.html"><a href="epigenetic-data.html#ucscs-genome-browser"><i class="fa fa-check"></i><b>1.4</b> UCSC’S Genome Browser</a></li>
<li class="chapter" data-level="1.5" data-path="epigenetic-data.html"><a href="epigenetic-data.html#data-loaders-and-simplifying-pre-processing"><i class="fa fa-check"></i><b>1.5</b> Data loaders and simplifying pre-processing</a></li>
<li class="chapter" data-level="1.6" data-path="epigenetic-data.html"><a href="epigenetic-data.html#dealing-with-missing-data-oversampling-undersampling-weighting"><i class="fa fa-check"></i><b>1.6</b> Dealing with missing data (oversampling, undersampling, weighting)</a></li>
</ul></li>
<li class="part"><span><b>II Training models with DNA input</b></span></li>
<li class="chapter" data-level="2" data-path="loss-functions-and-peak-metrics.html"><a href="loss-functions-and-peak-metrics.html"><i class="fa fa-check"></i><b>2</b> Loss functions, and peak metrics</a>
<ul>
<li class="chapter" data-level="2.1" data-path="loss-functions-and-peak-metrics.html"><a href="loss-functions-and-peak-metrics.html#reverse-complements-and-sequence-shifts"><i class="fa fa-check"></i><b>2.1</b> Reverse Complements and Sequence Shifts</a></li>
<li class="chapter" data-level="2.2" data-path="loss-functions-and-peak-metrics.html"><a href="loss-functions-and-peak-metrics.html#hyper-parameter-optimisation"><i class="fa fa-check"></i><b>2.2</b> Hyper-parameter optimisation</a></li>
</ul></li>
<li class="part"><span><b>III Software libraries for model building</b></span></li>
<li class="chapter" data-level="3" data-path="grelu.html"><a href="grelu.html"><i class="fa fa-check"></i><b>3</b> gReLU</a></li>
<li class="chapter" data-level="4" data-path="kipoi.html"><a href="kipoi.html"><i class="fa fa-check"></i><b>4</b> Kipoi</a></li>
<li class="chapter" data-level="5" data-path="weights-and-biases.html"><a href="weights-and-biases.html"><i class="fa fa-check"></i><b>5</b> Weights and Biases</a></li>
<li class="part"><span><b>IV ML pitfalls in genomics</b></span></li>
<li class="chapter" data-level="6" data-path="pitfalls-overview.html"><a href="pitfalls-overview.html"><i class="fa fa-check"></i><b>6</b> Pitfalls overview</a>
<ul>
<li class="chapter" data-level="6.1" data-path="pitfalls-overview.html"><a href="pitfalls-overview.html#distributional-differences"><i class="fa fa-check"></i><b>6.1</b> Distributional differences</a></li>
<li class="chapter" data-level="6.2" data-path="pitfalls-overview.html"><a href="pitfalls-overview.html#dependent-examples"><i class="fa fa-check"></i><b>6.2</b> Dependent examples</a></li>
<li class="chapter" data-level="6.3" data-path="pitfalls-overview.html"><a href="pitfalls-overview.html#confounding"><i class="fa fa-check"></i><b>6.3</b> Confounding</a></li>
<li class="chapter" data-level="6.4" data-path="pitfalls-overview.html"><a href="pitfalls-overview.html#leaky-pre-processing"><i class="fa fa-check"></i><b>6.4</b> Leaky pre-processing</a></li>
<li class="chapter" data-level="6.5" data-path="pitfalls-overview.html"><a href="pitfalls-overview.html#unbalanced-classes"><i class="fa fa-check"></i><b>6.5</b> Unbalanced classes</a></li>
</ul></li>
<li class="part"><span><b>V Model interpretability</b></span></li>
<li class="chapter" data-level="7" data-path="creating-and-visualising-a-simple-model.html"><a href="creating-and-visualising-a-simple-model.html"><i class="fa fa-check"></i><b>7</b> Creating and visualising a simple model</a></li>
<li class="chapter" data-level="8" data-path="tf-mo-disco.html"><a href="tf-mo-disco.html"><i class="fa fa-check"></i><b>8</b> TF mo-Disco</a></li>
<li class="part"><span><b>VI Using existing models</b></span></li>
<li class="chapter" data-level="9" data-path="using-the-grelu-model-zoo.html"><a href="using-the-grelu-model-zoo.html"><i class="fa fa-check"></i><b>9</b> Using the gReLU model zoo</a></li>
<li class="chapter" data-level="10" data-path="fine-tuning-of-enformer.html"><a href="fine-tuning-of-enformer.html"><i class="fa fa-check"></i><b>10</b> Fine tuning of Enformer</a></li>
<li class="part"><span><b>VII Predicting in novel cell types</b></span></li>
<li class="chapter" data-level="11" data-path="incorporating-atac-seq-info.html"><a href="incorporating-atac-seq-info.html"><i class="fa fa-check"></i><b>11</b> Incorporating ATAC-seq info</a></li>
<li class="chapter" data-level="12" data-path="use-of-cell-type-averages.html"><a href="use-of-cell-type-averages.html"><i class="fa fa-check"></i><b>12</b> Use of cell type averages</a></li>
<li class="part"><span><b>VIII More complex models</b></span></li>
<li class="chapter" data-level="13" data-path="training-multi-headed-models.html"><a href="training-multi-headed-models.html"><i class="fa fa-check"></i><b>13</b> Training multi-headed models</a></li>
<li class="chapter" data-level="14" data-path="training-siamese-twin-models.html"><a href="training-siamese-twin-models.html"><i class="fa fa-check"></i><b>14</b> Training siamese twin models</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Machine Learning in Genomics: Containerised tutorials demonstrating best practises, pitfalls, and reproducibility</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="grelu" class="section level1 hasAnchor" number="3">
<h1><span class="header-section-number">3</span> gReLU<a href="grelu.html#grelu" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>“<a href="https://github.com/Genentech/gReLU/blob/main/README.md" target="_blank">gReLU</a>” is a Python library to train, interpret, and apply deep learning models to DNA sequences”. The gRelu library contains a model zoo allowing for easy access to several models such as Borzoi, Enformer, or a dilated convolutional model based on the BPnet model architecture. Some of these models can be imported pre-trained. Additionally, simpler base models and convolutional neural networks are also available. On top of access to already built models, the software library allows for designing your own models.</p>
<img src="https://raw.githubusercontent.com/Genentech/gReLU/main/media/flowchart.jpg" />
<div style="text-align: center;">
<p>Source: <a href="https://github.com/Genentech/gReLU/blob/main/README.md" target="_blank">gRelu</a></p>
</div>
<p><br></p>
<p>Their <a href="https://genentech.github.io/gReLU/index.html" target="_blank">documentation</a> contains all their available functions and includes tutorials on using gRelu.</p>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="loss-functions-and-peak-metrics.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="kipoi.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/03-libraries-model-building.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
