<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>8 Pitfalls overview | Machine Learning in Genomics: Containerised tutorials demonstrating best practises, pitfalls, and reproducibility</title>
  <meta name="description" content="a set of reproducible, containerized tutorials that include all necessary data, code, and descriptions to replicate key results, along with demonstrations of common pitfalls in the field of genomics." />
  <meta name="generator" content="bookdown 0.39 and GitBook 2.6.7" />

  <meta property="og:title" content="8 Pitfalls overview | Machine Learning in Genomics: Containerised tutorials demonstrating best practises, pitfalls, and reproducibility" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="a set of reproducible, containerized tutorials that include all necessary data, code, and descriptions to replicate key results, along with demonstrations of common pitfalls in the field of genomics." />
  <meta name="github-repo" content="https://github.com/sn2023imperial/genomicsmarkdown" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="8 Pitfalls overview | Machine Learning in Genomics: Containerised tutorials demonstrating best practises, pitfalls, and reproducibility" />
  
  <meta name="twitter:description" content="a set of reproducible, containerized tutorials that include all necessary data, code, and descriptions to replicate key results, along with demonstrations of common pitfalls in the field of genomics." />
  

<meta name="author" content=" Sach Nehal" />


<meta name="date" content="2024-09-03" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="software-libraries-for-model-building.html"/>
<link rel="next" href="using-grelu-models.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>
<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Machine Learning in Genomics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>About</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="epigenetic-data.html"><a href="epigenetic-data.html"><i class="fa fa-check"></i><b>1</b> Epigenetic Data</a>
<ul>
<li class="chapter" data-level="1.1" data-path="epigenetic-data.html"><a href="epigenetic-data.html#what-is-epigenetic-data"><i class="fa fa-check"></i><b>1.1</b> What is epigenetic data?</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="epigenetic-data.html"><a href="epigenetic-data.html#what-does-dna-look-like"><i class="fa fa-check"></i><b>1.1.1</b> What Does DNA Look Like?</a></li>
<li class="chapter" data-level="1.1.2" data-path="epigenetic-data.html"><a href="epigenetic-data.html#common-epigenetic-sequencing-techniques"><i class="fa fa-check"></i><b>1.1.2</b> Common Epigenetic Sequencing Techniques:</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="epigenetic-data.html"><a href="epigenetic-data.html#what-does-epigenetic-data-look-like"><i class="fa fa-check"></i><b>1.2</b> What does epigenetic data look like?</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="epigenetic-data.html"><a href="epigenetic-data.html#representing-epigenetic-data"><i class="fa fa-check"></i><b>1.2.1</b> Representing epigenetic data</a></li>
<li class="chapter" data-level="1.2.2" data-path="epigenetic-data.html"><a href="epigenetic-data.html#transformations-to-stop-extreme-p-values"><i class="fa fa-check"></i><b>1.2.2</b> Transformations to stop extreme p-values</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="epigenetic-data.html"><a href="epigenetic-data.html#sources-of-epigenetic-data"><i class="fa fa-check"></i><b>1.3</b> Sources of epigenetic data</a></li>
<li class="chapter" data-level="1.4" data-path="epigenetic-data.html"><a href="epigenetic-data.html#ucscs-genome-browser"><i class="fa fa-check"></i><b>1.4</b> UCSC’S Genome Browser</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="pre-processing-of-bigwig-files.html"><a href="pre-processing-of-bigwig-files.html"><i class="fa fa-check"></i><b>2</b> Pre-processing of bigWig files</a>
<ul>
<li class="chapter" data-level="2.1" data-path="pre-processing-of-bigwig-files.html"><a href="pre-processing-of-bigwig-files.html#data-loaders-and-simplifying-pre-processing"><i class="fa fa-check"></i><b>2.1</b> Data loaders and simplifying pre-processing</a></li>
<li class="chapter" data-level="2.2" data-path="pre-processing-of-bigwig-files.html"><a href="pre-processing-of-bigwig-files.html#dealing-with-missing-data-oversampling-undersampling-weighting"><i class="fa fa-check"></i><b>2.2</b> Dealing with missing data (oversampling, undersampling, weighting)</a></li>
</ul></li>
<li class="part"><span><b>II Training models with DNA input</b></span></li>
<li class="chapter" data-level="3" data-path="loss-functions-and-peak-metrics.html"><a href="loss-functions-and-peak-metrics.html"><i class="fa fa-check"></i><b>3</b> Loss functions, and peak metrics</a></li>
<li class="chapter" data-level="4" data-path="training-tricks.html"><a href="training-tricks.html"><i class="fa fa-check"></i><b>4</b> Training tricks</a>
<ul>
<li class="chapter" data-level="4.1" data-path="training-tricks.html"><a href="training-tricks.html#reverse-complements-and-sequence-shifts"><i class="fa fa-check"></i><b>4.1</b> Reverse Complements and Sequence Shifts</a></li>
<li class="chapter" data-level="4.2" data-path="training-tricks.html"><a href="training-tricks.html#hyper-parameter-optimisation"><i class="fa fa-check"></i><b>4.2</b> Hyper-parameter optimisation</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="reproducibility-of-machine-learning-models.html"><a href="reproducibility-of-machine-learning-models.html"><i class="fa fa-check"></i><b>5</b> Reproducibility of machine learning models</a></li>
<li class="chapter" data-level="6" data-path="testing.html"><a href="testing.html"><i class="fa fa-check"></i><b>6</b> Testing</a></li>
<li class="chapter" data-level="7" data-path="software-libraries-for-model-building.html"><a href="software-libraries-for-model-building.html"><i class="fa fa-check"></i><b>7</b> Software libraries for model building</a>
<ul>
<li class="chapter" data-level="7.1" data-path="software-libraries-for-model-building.html"><a href="software-libraries-for-model-building.html#grelu"><i class="fa fa-check"></i><b>7.1</b> gReLU</a></li>
<li class="chapter" data-level="7.2" data-path="software-libraries-for-model-building.html"><a href="software-libraries-for-model-building.html#kipoi"><i class="fa fa-check"></i><b>7.2</b> Kipoi</a></li>
</ul></li>
<li class="part"><span><b>III ML pitfalls in genomics</b></span></li>
<li class="chapter" data-level="8" data-path="pitfalls-overview.html"><a href="pitfalls-overview.html"><i class="fa fa-check"></i><b>8</b> Pitfalls overview</a></li>
<li class="part"><span><b>IV Using existing models</b></span></li>
<li class="chapter" data-level="9" data-path="using-grelu-models.html"><a href="using-grelu-models.html"><i class="fa fa-check"></i><b>9</b> Using gReLU models</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Machine Learning in Genomics: Containerised tutorials demonstrating best practises, pitfalls, and reproducibility</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="pitfalls-overview" class="section level1 hasAnchor" number="8">
<h1><span class="header-section-number">8</span> Pitfalls overview<a href="pitfalls-overview.html#pitfalls-overview" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Applying machine learning in the field of genomics comes with its own challenges. Genomic data is highly complex, featuring high dimensionality, heterogeneity, and noise. It is important to consider the assumptions models make and whether those assumptions hold true within the data. According to research, available <a href="https://www.nature.com/articles/s41576-021-00434-9" target="_blank">here</a>, on pitfalls in machine learning, the most common errors include not taking into account distributional differences between contexts or batches, dependencies within the data, confounding effects distorting true relationships, unbalanced classes and leaking information between datasets.<a href="#fn45" class="footnote-ref" id="fnref45"><sup>45</sup></a> Lets examine the pitfalls mentioned in the research above and methods to mitigate these effects in more detail.</p>
<img src="images/pitfalls.png" />
<div style="text-align: center;">
<p>Source: <a href="https://www.nature.com/articles/s41576-021-00434-9/figures/1" target="_blank">Nature Reviews Genetics</a></p>
</div>
<p><br></p>
<p><strong>Distributional differences</strong></p>
<p>Distributional differences between the context in which a model was trained and tested, and the context in which a model makes predictions, results in different model performances. These different contexts can be attributed to batch effects or applying the model on different cell types. “One should expect that performance will be higher in [the same setting] than on [different settings]<a href="#fn46" class="footnote-ref" id="fnref46"><sup>46</sup></a>. It is crucial then, for models which aim to predict biological relationships across contexts to be validated and tested across these different settings that accurately reflect the real-world variability. Even models that aim to predict across one context such as cell type-specific models common in disease prediction, are still susceptible to this pitfall. When using data from multiple experiments, batch effects can introduce variability that lead to differences between training, test, and prediction sets. To address this, it is important to consider batch effects during model development and evaluation. One approach is to ensure that training and test sets are sourced from different batches to evaluate the generalisability of the model and prevent inflated performance. Additionally, applying batch effect correction techniques such can help account for these distributional differences.
<br><br></p>
<p><strong>Dependent examples</strong></p>
<p>When studying machine learning, a constant assumption is that of independence. Whether it be independent test sets or data points being independent of each other, the assumption of independence is key in creating fair models. Many biological processes in genomics are not independent.<a href="#fn47" class="footnote-ref" id="fnref47"><sup>47</sup></a> A research paper focused on understanding the pitfalls of neural networks predicting across cell types found that several models had been evaluated on test sets comprising of cell type independent of training, but not independent of chromosomes leading to inflated performances.<a href="#fn48" class="footnote-ref" id="fnref48"><sup>48</sup></a> That is, gene expression predictions were evaluated on different cells but on the same chromosomes. The inflated performances were due to chromosomes “themselves [being] dependent across samples because the underlying functional activity is generally shared”<a href="#fn49" class="footnote-ref" id="fnref49"><sup>49</sup></a>. While hard to identify in the designing phase even through visualisation, mitigation techniques include preventing overfitting during evaluation and group k-fold cross-validation to prevent leakage between training and test sets.<a href="#fn50" class="footnote-ref" id="fnref50"><sup>50</sup></a>
<br><br></p>
<p><strong>Confounding</strong></p>
<p>Confounders are additional variables that affect the variables being studied, resulting in models not capturing the correct relationships in the data.<a href="#fn51" class="footnote-ref" id="fnref51"><sup>51</sup></a> “Confounding in genetic studies can arise from unmodelled environmental factors and population structure, as well as other factors”.<a href="#fn52" class="footnote-ref" id="fnref52"><sup>52</sup></a> In the context of ATAC-Seq and chIP-Seq data, differences in the sequencing depth of the experiment can have a confounding effect on models if not handled. As explained in part 1, ATAC-Seq and chIP-Seq data consists of reads that are alligned to a reference genome and aggregated. The sequencing depth of the experiment refers to on average how many times a region was sequenced.<a href="#fn53" class="footnote-ref" id="fnref53"><sup>53</sup></a> A higher sequencing depth means a higher overall coverage level. If not accounted for, models using data of different sequencing depths are confounded. Similarly when predicting on a dataset with a different sequencing depth than the training set, the model would biased. While sequencing depth as a confounder can be more easily corrected by downsampling the raw reads from the dataset with higher sequencing depth, other confounding effects can be harder to account for. However, it is possible to use statistical tools such as PEER<a href="#fn54" class="footnote-ref" id="fnref54"><sup>54</sup></a>, Inter-sample Correlation Emended (ICE) and Surrogate Variable Analysis (SVA)<a href="#fn55" class="footnote-ref" id="fnref55"><sup>55</sup></a> to understand confounders in your dataset.
<br><br></p>
<p><strong>Leaky pre-processing</strong></p>
<p>Leaky pre-processing involves information leaks between the training and test sets resulting in altered testing metrics. In genomics, as a result of many dependencies, pre-processing steps involving the dataset as a whole can introduce this bias. This includes standardisation techniques, supervised feature selection or principle component analysis, before splitting data into test splits.<a href="#fn56" class="footnote-ref" id="fnref56"><sup>56</sup></a> A solution to this is to perform these transformations and analysis after data splits, preferably within cross-validation. Leaky preprocessing has been prevalent in various genomic fields, including microarray analysis, DNA methylation, gene expression studies, and more, often leading to misleading results.<a href="#fn57" class="footnote-ref" id="fnref57"><sup>57</sup></a>
<br><br></p>
<p><strong>Unbalanced classes</strong></p>
<p>Unbalanced datasets can be common in genomics often necessitating the use of statistical methods to validate the few positive instances amid vast amounts of data. In these cases, the pitfall is a model “learns most of the target concepts of the majority class and learns target concepts from the minority class poorly or not at all.”<a href="#fn58" class="footnote-ref" id="fnref58"><sup>58</sup></a> This is particularly evident in disease related tasks where the focus is on a few disease causing genes or non-coding variants.<a href="#fn59" class="footnote-ref" id="fnref59"><sup>59</sup></a> Additionally, many experiments that implement conservative significance thresholds to determine true signals involve data imbalance, such as peak detection<a href="#fn60" class="footnote-ref" id="fnref60"><sup>60</sup></a>, gene expression<a href="#fn61" class="footnote-ref" id="fnref61"><sup>61</sup></a>, and chromatin accessibility. The extensive size of the human genome exacerbates this issue when the areas of interest are small. Researchers address this imbalance by employing balancing algorithms that oversample the negative class and undersample the majority class. For instance, in training models to predict functional peaks from ChIP-seq or chromatin accessibility data, an approach might involve using all identified peaks along with a matching number of negative regions, thus effectively undersampling the majority class. For datasets with no such negative regions, researchers have to construct their own. While such imbalances are commonly discussed in classification, they also pose challenges in regression models predicting quantitative outcomes, where performance may be compromised in regions with sparse data, such as genomic areas or genes with low read counts in single-cell genomics studies.<a href="#fn62" class="footnote-ref" id="fnref62"><sup>62</sup></a></p>


</div>



<h3>References<a href="references.html#references" class="anchor-section" aria-label="Anchor link to header"></a></h3>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-allis2015" class="csl-entry">
Allis, D., Jenuwein T, Reinberg D, and Caparros M. L. 2015. <em>EPIGENETICS</em>. 2nd ed. Cold Spring Harbor Laboratory Press. <a href="https://www.cshlpress.com/pdf/sample/2014/epigenetics2/EPIFM.pdf">https://www.cshlpress.com/pdf/sample/2014/epigenetics2/EPIFM.pdf</a>.
</div>
<div id="ref-avsec2021" class="csl-entry">
Avsec, Žiga, Vikram Agarwal, Daniel Visentin, and et al. 2021. <em>Effective Gene Expression Prediction from Sequence by Integrating Long-Range Interactions</em>. Springer Nature. <a href="https://www.nature.com/articles/s41592-021-01252-x">https://www.nature.com/articles/s41592-021-01252-x</a>.
</div>
<div id="ref-muk2014" class="csl-entry">
Haque, Muksitul, and et al. 2014. <em>Imbalanced Class Learning in Epigenetics</em>. Journal of Computational Biology. <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4082351/">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4082351/</a>.
</div>
<div id="ref-jen2010" class="csl-entry">
Listgarten, Jennifer, and et al. 2010. <em>Correction for Hidden Confounders in the Genetic Analysis of Gene Expression</em>. Proceedings of the National Academy of Sciences of the United States of America. <a href="https://www.pnas.org/doi/full/10.1073/pnas.1002425107">https://www.pnas.org/doi/full/10.1073/pnas.1002425107</a>.
</div>
<div id="ref-oh2020" class="csl-entry">
Oh, Dongpin, Seth Strattan, Junho Hur, and et al. 2020. <em>CNN-Peaks: ChIP-Seq Peak Detection Pipeline Using Convolutional Neural Networks That Imitate Human Visual Inspection</em>. Springer Nature. <a href="https://www.nature.com/articles/s41598-020-64655-4">https://www.nature.com/articles/s41598-020-64655-4</a>.
</div>
<div id="ref-mo2012" class="csl-entry">
Pourhoseingholi, Mohamad, and et al. 2012. <em>How to Control Confounding Effects by Statistical Analysis</em>. Gastroenterol Hepatol Bed Bench. <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4017459/#:~:text=A%20Confounder%20is%20an%20extraneous,between%20the%20variables%20under%20study.">https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4017459/#:~:text=A%20Confounder%20is%20an%20extraneous,between%20the%20variables%20under%20study.</a>
</div>
<div id="ref-jacob2020" class="csl-entry">
Schreiber, Jacob, and et al. 2020. <em>A Pitfall for Machine Learning Methods Aiming to Predict Across Cell Types</em>. Springer Nature. <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02177-y">https://genomebiology.biomedcentral.com/articles/10.1186/s13059-020-02177-y</a>.
</div>
<div id="ref-sims2014" class="csl-entry">
Sims, David, and et al. 2014. <em>Sequencing Depth and Coverage: Key Considerations in Genomic Analyses</em>. Nature Reviews Genetics. <a href="https://www.nature.com/articles/nrg3642">https://www.nature.com/articles/nrg3642</a>.
</div>
<div id="ref-oliver2012" class="csl-entry">
Stegle, Oliver, and et al. 2012. <em>Using Probabilistic Estimation of Expression Residuals (PEER) to Obtain Increased Power and Interpretability of Gene Expression Analyses</em>. Nature Protocols. <a href="https://www.nature.com/articles/nprot.2011.457">https://www.nature.com/articles/nprot.2011.457</a>.
</div>
<div id="ref-whalen2022" class="csl-entry">
Whalen, Sean, Jacob Schreiber, William Noble, and Katherine Pollard. 2022. <span>“Navigating the Pitfalls of Applying Machine Learning in Genomics.”</span> <a href="https://www.nature.com/articles/s41576-021-00434-9">https://www.nature.com/articles/s41576-021-00434-9</a>.
</div>
</div>
<div class="footnotes">
<hr />
<ol start="45">
<li id="fn45"><p><span class="citation">Whalen et al. (<a href="#ref-whalen2022">2022</a>)</span><a href="pitfalls-overview.html#fnref45" class="footnote-back">↩︎</a></p></li>
<li id="fn46"><p><span class="citation">Whalen et al. (<a href="#ref-whalen2022">2022</a>)</span><a href="pitfalls-overview.html#fnref46" class="footnote-back">↩︎</a></p></li>
<li id="fn47"><p><span class="citation">Allis et al. (<a href="#ref-allis2015">2015</a>)</span><a href="pitfalls-overview.html#fnref47" class="footnote-back">↩︎</a></p></li>
<li id="fn48"><p><span class="citation">Schreiber and al (<a href="#ref-jacob2020">2020</a>)</span><a href="pitfalls-overview.html#fnref48" class="footnote-back">↩︎</a></p></li>
<li id="fn49"><p><span class="citation">Whalen et al. (<a href="#ref-whalen2022">2022</a>)</span><a href="pitfalls-overview.html#fnref49" class="footnote-back">↩︎</a></p></li>
<li id="fn50"><p><span class="citation">Whalen et al. (<a href="#ref-whalen2022">2022</a>)</span><a href="pitfalls-overview.html#fnref50" class="footnote-back">↩︎</a></p></li>
<li id="fn51"><p><span class="citation">Pourhoseingholi and al (<a href="#ref-mo2012">2012</a>)</span><a href="pitfalls-overview.html#fnref51" class="footnote-back">↩︎</a></p></li>
<li id="fn52"><p><span class="citation">Whalen et al. (<a href="#ref-whalen2022">2022</a>)</span><a href="pitfalls-overview.html#fnref52" class="footnote-back">↩︎</a></p></li>
<li id="fn53"><p><span class="citation">Sims and al (<a href="#ref-sims2014">2014</a>)</span><a href="pitfalls-overview.html#fnref53" class="footnote-back">↩︎</a></p></li>
<li id="fn54"><p><span class="citation">Stegle and al (<a href="#ref-oliver2012">2012</a>)</span><a href="pitfalls-overview.html#fnref54" class="footnote-back">↩︎</a></p></li>
<li id="fn55"><p><span class="citation">Listgarten and al (<a href="#ref-jen2010">2010</a>)</span><a href="pitfalls-overview.html#fnref55" class="footnote-back">↩︎</a></p></li>
<li id="fn56"><p><span class="citation">Whalen et al. (<a href="#ref-whalen2022">2022</a>)</span><a href="pitfalls-overview.html#fnref56" class="footnote-back">↩︎</a></p></li>
<li id="fn57"><p><span class="citation">Whalen et al. (<a href="#ref-whalen2022">2022</a>)</span><a href="pitfalls-overview.html#fnref57" class="footnote-back">↩︎</a></p></li>
<li id="fn58"><p><span class="citation">Haque and al (<a href="#ref-muk2014">2014</a>)</span><a href="pitfalls-overview.html#fnref58" class="footnote-back">↩︎</a></p></li>
<li id="fn59"><p><span class="citation">Haque and al (<a href="#ref-muk2014">2014</a>)</span><a href="pitfalls-overview.html#fnref59" class="footnote-back">↩︎</a></p></li>
<li id="fn60"><p><span class="citation">Oh et al. (<a href="#ref-oh2020">2020</a>)</span><a href="pitfalls-overview.html#fnref60" class="footnote-back">↩︎</a></p></li>
<li id="fn61"><p><span class="citation">Avsec et al. (<a href="#ref-avsec2021">2021</a>)</span><a href="pitfalls-overview.html#fnref61" class="footnote-back">↩︎</a></p></li>
<li id="fn62"><p><span class="citation">Whalen et al. (<a href="#ref-whalen2022">2022</a>)</span><a href="pitfalls-overview.html#fnref62" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="software-libraries-for-model-building.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="using-grelu-models.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/04-pitfalls.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.9/latest.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
