<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>5 Using gReLU models to examine pitfalls | Machine Learning in Genomics: Containerised tutorials demonstrating best practises, pitfalls, and reproducibility</title>
  <meta name="description" content="a set of reproducible, containerized tutorials that include all necessary data, code, and descriptions to replicate key results, along with demonstrations of common pitfalls in the field of genomics." />
  <meta name="generator" content="bookdown 0.39 and GitBook 2.6.7" />

  <meta property="og:title" content="5 Using gReLU models to examine pitfalls | Machine Learning in Genomics: Containerised tutorials demonstrating best practises, pitfalls, and reproducibility" />
  <meta property="og:type" content="book" />
  
  <meta property="og:description" content="a set of reproducible, containerized tutorials that include all necessary data, code, and descriptions to replicate key results, along with demonstrations of common pitfalls in the field of genomics." />
  <meta name="github-repo" content="https://github.com/sn2023imperial/genomicsmarkdown" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="5 Using gReLU models to examine pitfalls | Machine Learning in Genomics: Containerised tutorials demonstrating best practises, pitfalls, and reproducibility" />
  
  <meta name="twitter:description" content="a set of reproducible, containerized tutorials that include all necessary data, code, and descriptions to replicate key results, along with demonstrations of common pitfalls in the field of genomics." />
  

<meta name="author" content=" Sach Nehal" />


<meta name="date" content="2024-09-07" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="pitfalls-overview.html"/>
<link rel="next" href="references.html"/>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections-1.1.0/anchor-sections.css" rel="stylesheet" />
<link href="libs/anchor-sections-1.1.0/anchor-sections-hash.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.1.0/anchor-sections.js"></script>



<style type="text/css">
  
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Machine Learning in Genomics</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>About</a></li>
<li class="part"><span><b>I Introduction</b></span></li>
<li class="chapter" data-level="1" data-path="epigenetic-data.html"><a href="epigenetic-data.html"><i class="fa fa-check"></i><b>1</b> Epigenetic Data</a>
<ul>
<li class="chapter" data-level="1.1" data-path="epigenetic-data.html"><a href="epigenetic-data.html#what-is-epigenetic-data"><i class="fa fa-check"></i><b>1.1</b> What is epigenetic data?</a>
<ul>
<li class="chapter" data-level="1.1.1" data-path="epigenetic-data.html"><a href="epigenetic-data.html#what-does-dna-look-like"><i class="fa fa-check"></i><b>1.1.1</b> What Does DNA Look Like?</a></li>
<li class="chapter" data-level="1.1.2" data-path="epigenetic-data.html"><a href="epigenetic-data.html#common-epigenetic-sequencing-techniques"><i class="fa fa-check"></i><b>1.1.2</b> Common Epigenetic Sequencing Techniques:</a></li>
</ul></li>
<li class="chapter" data-level="1.2" data-path="epigenetic-data.html"><a href="epigenetic-data.html#what-does-epigenetic-data-look-like"><i class="fa fa-check"></i><b>1.2</b> What does epigenetic data look like?</a>
<ul>
<li class="chapter" data-level="1.2.1" data-path="epigenetic-data.html"><a href="epigenetic-data.html#representing-epigenetic-data"><i class="fa fa-check"></i><b>1.2.1</b> Representing epigenetic data</a></li>
</ul></li>
<li class="chapter" data-level="1.3" data-path="epigenetic-data.html"><a href="epigenetic-data.html#sources-of-epigenetic-data"><i class="fa fa-check"></i><b>1.3</b> Sources of epigenetic data</a></li>
<li class="chapter" data-level="1.4" data-path="epigenetic-data.html"><a href="epigenetic-data.html#ucscs-genome-browser"><i class="fa fa-check"></i><b>1.4</b> UCSC’S Genome Browser</a>
<ul>
<li class="chapter" data-level="1.4.1" data-path="epigenetic-data.html"><a href="epigenetic-data.html#transformations-to-stop-extreme-p-values"><i class="fa fa-check"></i><b>1.4.1</b> Transformations to stop extreme p-values</a></li>
</ul></li>
<li class="chapter" data-level="1.5" data-path="epigenetic-data.html"><a href="epigenetic-data.html#data-loaders-and-simplifying-pre-processing"><i class="fa fa-check"></i><b>1.5</b> Data loaders and simplifying pre-processing</a></li>
</ul></li>
<li class="part"><span><b>II Training models with DNA input</b></span></li>
<li class="chapter" data-level="2" data-path="loss-functions.html"><a href="loss-functions.html"><i class="fa fa-check"></i><b>2</b> Loss functions</a>
<ul>
<li class="chapter" data-level="2.1" data-path="loss-functions.html"><a href="loss-functions.html#reverse-complements-and-sequence-shifts"><i class="fa fa-check"></i><b>2.1</b> Reverse Complements and Sequence Shifts</a></li>
<li class="chapter" data-level="2.2" data-path="loss-functions.html"><a href="loss-functions.html#hyper-parameter-optimisation"><i class="fa fa-check"></i><b>2.2</b> Hyper-parameter optimisation</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="software-libraries-for-model-building.html"><a href="software-libraries-for-model-building.html"><i class="fa fa-check"></i><b>3</b> Software libraries for model building</a>
<ul>
<li class="chapter" data-level="3.1" data-path="software-libraries-for-model-building.html"><a href="software-libraries-for-model-building.html#grelu"><i class="fa fa-check"></i><b>3.1</b> gReLU</a></li>
<li class="chapter" data-level="3.2" data-path="software-libraries-for-model-building.html"><a href="software-libraries-for-model-building.html#kipoi"><i class="fa fa-check"></i><b>3.2</b> Kipoi</a></li>
</ul></li>
<li class="part"><span><b>III ML pitfalls in genomics</b></span></li>
<li class="chapter" data-level="4" data-path="pitfalls-overview.html"><a href="pitfalls-overview.html"><i class="fa fa-check"></i><b>4</b> Pitfalls overview</a></li>
<li class="part"><span><b>IV Using existing models</b></span></li>
<li class="chapter" data-level="5" data-path="using-grelu-models-to-examine-pitfalls.html"><a href="using-grelu-models-to-examine-pitfalls.html"><i class="fa fa-check"></i><b>5</b> Using gReLU models to examine pitfalls</a></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Machine Learning in Genomics: Containerised tutorials demonstrating best practises, pitfalls, and reproducibility</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="using-grelu-models-to-examine-pitfalls" class="section level1 hasAnchor" number="5">
<h1><span class="header-section-number">5</span> Using gReLU models to examine pitfalls<a href="using-grelu-models-to-examine-pitfalls.html#using-grelu-models-to-examine-pitfalls" class="anchor-section" aria-label="Anchor link to header"></a></h1>
<p>Continuing from previous tutorials, the next tutorial similarly uses chIP-Seq data from Encode Experiment <a href="https://www.encodeproject.org/experiments/ENCSR817LUF/" target="_blank">ENCSR817LUF</a>. In tutorial 2 we use this data to train a gRelu convolutional neural network in order to to predict the coverage levels and examine several of the pitfalls mentioned in section 4. An additional tutorial (tutorial1.5) is included to explain how data was pre-processed for tutorial 2. Reading through gRelu’s own <a href="https://genentech.github.io/gReLU/tutorials/3_train.html#" target="_blank">tutorial</a> on training their models is recommended.</p>
<p><a href="https://colab.research.google.com/drive/1wWlFffwBpOuK9yyCWp1hY6oiG9FK_6aX?usp=sharing" target="_blank">Tutorial 1.5: Preprocessing Data for gRelu Models (interactive)</a></p>
<p><a href="https://nbviewer.org/github/sn2023imperial/genomicsmarkdown/blob/main/notebook_tutorials/Tutorial_1.5_Preprocessing_gRelu/Tutorial1.5_Preprocessing_gRelu.ipynb" target="_blank">Tutorial 1.5: Preprocessing Data for gRelu Models (nbviewer)</a>
<br></p>
<p><a href="https://colab.research.google.com/drive/17xdEjl6TDew8Fb-bZvjRMIQSZX5s95Ae?usp=sharing" target="_blank">Tutorial 2: Training Models with gRelu and Examining Pitfalls (interactive)</a></p>
<p><a href="https://nbviewer.org/github/sn2023imperial/genomicsmarkdown/blob/main/notebook_tutorials/Tutorial_2_Examining_Pitfalls_with_gRelu/Tutorial2_Grelu.ipynb" target="_blank">Tutorial 2: Training Models with gRelu and Examining Pitfalls (nbviewer)</a>
<br></p>



</div>
            </section>

          </div>
        </div>
      </div>
<a href="pitfalls-overview.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="references.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/USERNAME/REPO/edit/BRANCH/06-using-existing-models.Rmd",
"text": "Edit"
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["_main.pdf", "_main.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

</body>

</html>
